### ISSUE TITLE
OpenID: Could not authenticate against third party

### ISSUE THREAD

**Issue by aleprovencio** [go-vikunja/desktop#12]
Hi!

I can't manage to connect `vikunja-desktop` to my instance behind `authelia`.

I do get correctly redirected to `authelia`'s `URL`, put my credentials, but then get a `Could not authenticate against third party` message on this app's `UI`.

This happens on the `flatpak` and `AUR` versions of `vikunja-desktop`. For the sake of testing, I installed `vikunja-app` on my phone, and could successfully authenticate.

`Authelia` logs show this error:
```
"Access Request failed with error: The provided authorization grant (e.g., authorization code, resource owner credentials) or refresh token is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client. The 'redirect_uri' from this request does not match the one from the authorize request." method=POST path=/api/oidc/token remote_ip=172.19.0.7
```

Relevant settings:

authelia:
```yaml

auth:
  local:
    enabled: true
  openid:
    enabled: true
    redirecturl: https://vikunja.{{ domain_name }}/auth/openid/
    providers:
      - name: Authelia
        authurl: https://authelia.{{ domain_name }}
        clientid: vikunja
        clientsecret: '{{ stack_apps_vikunja_openid_clientsecret }}'
```

vikunja:
```yaml
  - id: vikunja
    description: Vikunja
    secret: '{{ stack_proxy_auth_authelia_vikunja_secret }}'
    public: false
    authorization_policy: two_factor
    consent_mode: implicit
    redirect_uris:
      - https://vikunja.{{ domain_name }}/auth/openid/authelia
      - http://127.0.0.1:45735/auth/openid/authelia  # vikunja desktop
    scopes:
      - openid
      - email
      - profile
    userinfo_signing_algorithm: none
```

**Comment by kolaente** [comment]
If you start the desktop app and open `http://127.0.0.1:45735` in a browser, do you see the frontend?

**Comment by aleprovencio** [comment]
Yes, same thing happens on the browser and I can spot this error:

`Failed to load resource: the server responded with a status of 400 ()`

**Comment by kolaente** [comment]
Anything in Vikunja's api logs?

**Comment by aleprovencio** [comment]
Sorry, I missed that one, here you go:

```log
2023-12-15T16:45:47.474259218-03:00 2023-12-15T16:45:47.473204496-03:00: ERROR  ▶ openid/HandleCallback cb6 oauth2: "invalid_grant" "The provided authorization grant (e.g. , authorization code, resource owner credentials) or refresh token is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or w as issued to another client. The 'redirect_uri' from this request does not match the one from the authorize request."

2023-12-15T16:45:47.474320257-03:00 2023-12-15T16:45:47.473564408-03:00: WEB    ▶ 192.168.55.51  POST 400 /api/v1/auth/openid/authelia/callback 456.59392ms - Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/6.6.1 Chrome/112.0.5615.213 Safari/537.36
```

**Comment by kolaente** [comment]
The desktop app may change the port when there is another application already using it. Is it possible to use wildcards in Authelia's config to allow everything that's on `localhost` or `127.0.0.1`?

**Comment by aleprovencio** [comment]
I'm afraid `authelia` won't do it:
https://github.com/authelia/authelia/discussions/4659
https://www.authelia.com/configuration/identity-providers/open-id-connect/

However I can assure that the client is running on that port according to the output of `flatpak run io.vikunja.Vikunja`

**Comment by kolaente** [comment]
And it works fine with the Vikunja frontend you have set up on the web?

**Comment by aleprovencio** [comment]
Yes, the frontend works fine and also `vikunja-app` authenticates with no issues.

**Comment by kolaente** [comment]
I have an idea where the problem might be. Can you set the `redirecturl` in Vikunja's api config to `http://127.0.0.1:45735/auth/openid/authelia` and try again?

**Comment by aleprovencio** [comment]
Good one, now I see a different error on the browser:
```
Access to XMLHttpRequest at 'https://vikunja.{{ domain_name }}/api/v1/auth/openid/authelia/callback' from origin 'http://127.0.0.1:45735' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

Both `vikunja-api` and `authelia` seem OK.

I trying to add `http://127.0.0.1` on `authelia`'s `oidc.cors.allowed_origins` and `traefik`'s `headers.accessControlAllowOriginList`, still without success.

**Comment by kolaente** [comment]
This looks definitely like a bug. I'll take a look.

**Comment by kolaente** [comment]
Should be fixed in [`ee980e2a00`](https://kolaente.dev/vikunja/frontend/commit/ee980e2a002d6d57bc3d968a149b57c10ebac845) and [`ce53663a88`](https://kolaente.dev/vikunja/api/commit/ce53663a8853cc00d0da8c56ade77bfb522e335e). Please check with the next unstable build of the desktop and api (should be ready in ~1h) if the problem really went away.

**Comment by aleprovencio** [comment]
It is fixed indeed!

Unstable api release is enough, it works with the desktop's current version on flathub.

Thank you @kolaente !

**Comment by kolaente** [comment]
You're probably using the latest release then, 0.22.1, which was released yesterday.
